### PR TITLE
Buffer types for array backend

### PR DESCRIPTION
**Pull request by Tokazama**
## Motivation

After a conversation on slack with @StefanKarpinski and @brenhinkeller it seemed like a lot of other PRs on improving arrays stalled out and the first fundamental step is having a generic buffer type. I still have crashes in the REPL with this current implementation and it appears I haven't fully implemented the GC side correctly, but I figured it was better to put this out there for someone else to start with then completely abandon this.

The goal here is to provide the core functionality for bare-bones storage of multiple repeated elements.

## Implementation

There are three primary storage types this PR is intended to support:

* `Buffer{T}`: has a fixed size and mutable elements

* `DynamicBuffer{T}`: has a dynamic size and mutable elements.

* `ImmutableBuffer{T}`: has a fixed size and immutable elements (not implemented here but should be as a future PR for with the `freeze`/`thaw` compiler optimizations worked on in prior `ImmutableArray` PRs).

The type layout is modeled after that of `jl_array_t` and is similar to the following native Julia structure:

```julia
struct MemoryChunk{T}
    length::Int
    data::Ptr{T}
end
```

Just like `Array`, the number of bytes stored determines if `data` points to inline allocations that extends the size of `MemoryChunk` via `jl_gc_alloc` or a seperately stored chunk via `jl_gc_managed_malloc` and `jl_gc_track_malloced_buffer`.

Storage of elements is similar as bit unions, bits, or pointers to boxed types is identical to that of `Array` or `jl_array_t`.

Note that there are not explicitly stored flags, offsets, additional dimensions, or element size. This means that:

* Unlike `jl_array_t`, all information about the element type must be derived again everytime it's needed when running C code. Of course, this is not an issue once we are working with TBAA and that is all optimized away. However, I'm uncertain whether that will slow things down that don't often interact with TBAA directly (such as the garbage collector).

* We don't have `flags.how == 1` to mark a julia-allocated buffer that needs to be marked or `flags.isshared`. I've tried to use the last two its of the data pointer to mark these (see `jl_buffer_isshared` and `jl_buffer_isunmarked`). It may make more sense to have a type that explicitly wraps a shared buffer, preserving that bit for some other future use and allowing more explicit representation of data storage through the type system. I've yet to spend much time grocking the unmarked allocated data aspect of this, so there may be a better approach that I've yet to consider.

* The lack of an offset or maximum size means that resizing for `DynamicBuffer` will always result in a reallocation or new allocation. Feature parity with `Vector` is intended to be accomplished through Julia code with something like
    ```julia
    mutable struct DynamicVector{T} <: DenseVector{T}
        buffer::DynamicBuffer{T}
        offset::Int
        length::Int
    end
    ```
    where the length of `DynamicBuffer` here is functionally equivalent to `Array`s maxsize.

I'm in the process of moving the implementation to native Julia. This is a bit challenging since this is essentially an attempt to rewrite large portions of "array.c" where a lot of things aren't implemented in native Julia code (interacting with the GC, traversing `Union`s) and sometimes working efficiently with a pointer get tricky (storing pointers to immutable types).

## Remaining Work

* Names may not be sufficiently self documenting and may even be misleading. The term "buffer" is often used to refer to something more akin to what our current `Vector` is with offsets and resizing. Perhaps something that is more clearly describing a continuous chunk of memory such as `MemoryChunk`, `ResizableMemoryChunk`, and `ImmutableMemoryChunk`.

* What needs to be done here to support better allocation practices? There's a lot of interest in providing users with the ability specify how things are allocated (allocating to the stack, bump allocators, smart pointers). I've assumed that most of that would be better addressed in future PRs by those who really understand how to optimize that sort of thing, but I'd also like to ensure that the implementation here is not prohibittive to future developments.

* Could we provide better support for `Bool` here so that we don't need an explicit `BitBuffer` type?

* More support for future implementation of `DynamicVector` with user friendly resizing methods

* The assumed effects for methods should change based on the buffer variant. This is currently buggy. For example, the effects for `length(::Buffer)` should not be the same as `length(::DynamicBuffer)`, but currently are.

* Would it make sense to implement `ImmutableBuffer` here without any of the `thaw`/`freeze` stuff that would complicate things?

### CODE REVIEW

**Comment by quinnj** [comment]
Ha, @oscardssmith and I were just chatting about this about a month ago and we revived a branch I had started a couple years ago doing something similar: https://github.com/JuliaLang/julia/tree/add-buffer-type2. It takes a slightly different approach in basically copying `jl_array_t` (allowing any eltype) but having it be fixed size. We got it mostly working, but there were still a few GC-related pieces to fix.

(not saying this PR isn't good/right; just wanted to raise the fact that a parallel effort was restarted recently)

**Comment by oscardssmith** [comment]
@Tokazama I think a mix between our approaches is probably best. Specifically, I like that our version let you have a non UInt8 eltype, but I like your data layout for the struct a lot better.

**Comment by Tokazama** [comment]
I think allowing the eltype to be included is a good idea if we can keep the structure itself the same. This does complicate things quite a bit (particularly with the garbage collector) because you have to figure out size and alignment manually a lot more without fields that explicitly reference that. I wasn't confident enough in everything else I was doing to spend more time implementing that too.

**Comment by oscardssmith** [comment]
I think all you need to do is make a type parameter for the struct that stores the type, and then you can just get the size information from the type parameter.

**Comment by Tokazama** [comment]
If someone figured out the shared pointer thing so we could have a type that is truly transparent then it should be fairly simple. We'd basically have three pipelines for allocation:

1. `SimplBuffer{BitType}`
2. `SimpleBuffer{UnionBitType}`
3. `SimpleBuffer{SharedPtr{BigType}}`

Then we could have a `ThinArray` type that wraps it, hiding some of the ugliness from users without losing the transparency for other use cases.

> I think all you need to do is make a type parameter for the struct that stores the type, and then you can just get the size information from the type parameter.

That would definitely work, but since I was still getting errors with a very basic implementation I wasn't sure it was a good idea to add more complexity yet.

**Comment by Tokazama** [comment]
While it's on my mind I'll change around a couple things to make the transition to different types more approachable.

**Comment by quinnj** [comment]
I might not quite understand the details well enough, but I thought @vtjnash was against doing any kind of shared pointer exposure in a buffer type; `jl_value_t`s should only ever live in `/src` code and never exposed on the Julia side.

**Comment by Tokazama** [comment]
latest has a field for the type and the element size. I think I fixed all the code that is dependent on how this changes the amount of memory consumed but I could be wrong. I haven't had errors in my naive tests in repl but once we have settled on some other design decisions we clearly need tests here.

One of the current assumptions here is that `SimpleBuffer` always owns its data. So there's no pointer to another one containing the actual data. I prefer that this sort of transparency exists at this level, but I figured it needed to be explicitly pointed out given how this is different than `Array`.

**Comment by oscardssmith** [comment]
Is it really a good idea to add the immutable buffer in this PR? To me that seems like a lot of complexity that's mostly orthogonal.

**Comment by vtjnash** [comment]
There is, however, very little chance of this getting a proper review and approval if it contains both immutable and mutable buffer implementations simultaneously. Focus on one thing at a time. It is already questionable enough to decide if the breaking change of not supporting foreign pointers is going to be okay. Currently it appears to me that this implementation is not quite managing alignment correctly. I think we cannot have the `Buffer{T}` type simply always owning the data itself for these reasons, but rather require inventing a ownership token that will permit doing object inlining of a fat pointer pair `struct {gcptr, dataptr}`, where `gcptr.length` is the allocated length and `gcptr.data` may have same value as `dataptr` (but an extra pointer indirection away from Buffer), and frequently (not aways) we then have `gcptr + 16 == dataptr`. When they are separate allocations, we may need to attach a special finalizer operation too to call `free` when needed. The callee can choose to change dataptr to be an offset (view) into any part of the parent's `[gc.ptr, +gcptr.length+1]` bounds.

**Comment by oscardssmith** [comment]
All this complexity is why I really don't like having the separate pointer. The extra size is really annoying for uses that don't need resizing.

**Comment by quinnj** [comment]
> All this complexity is why I really don't like having the separate pointer. The extra size is really annoying for uses that don't need resizing.

I might be wrong, but I don't think @vtjnash is talking about an extra length to track size for resizing, but just the difference between the _allocated_ ptr size vs. user-facing size, with the only difference being alignment.

**Comment by oscardssmith** [comment]
iiuc we can get both of those from just looking at the type parameter.

**Comment by Tokazama** [comment]
> There is, however, very little chance of this getting a proper review and approval if it contains both immutable and mutable buffer implementations simultaneously. Focus on one thing at a time. It is already questionable enough to decide if the breaking change of not supporting foreign pointers is going to be okay. Currently it appears to me that this implementation is not quite managing alignment correctly. I think we cannot have the `Buffer{T}` type simply always owning the data itself for these reasons, but rather require inventing a ownership token that will permit doing object inlining of a fat pointer pair `struct {gcptr, dataptr}`, where `gcptr.length` is the allocated length and `gcptr.data` may have same value as `dataptr` (but an extra pointer indirection away from Buffer), and frequently (not aways) we then have `gcptr + 16 == dataptr`. When they are separate allocations, we may need to attach a special finalizer operation too to call `free` when needed. The callee can choose to change dataptr to be an offset (view) into any part of the parent's `[gc.ptr, +gcptr.length+1]` bounds.

There are two things that I think Julia is sorely missing that I want from this, transparent ownership and bare bones arrays. I think this will require at least three buffer types (mutable, immutable, and dynamic/resiable). I'm fine with doing one at a time in a PR but I wanted to make sure there was at least some perspective here that the solution is going to require more than just this.

One obvious thing the current implementation lacks is proper malloc tracked data. That can be done with the same type we have but instead of `jl_buffer_data(x)` returning a pointer to first element in the buffer it would return a pointer to the malloced buffer (same as it's done with `jl_array_t`.

I also haven't been able to fully figure out how to do the IR for this in "codegen.cpp", so performance is still pretty bad.

**Comment by Tokazama** [comment]
@vtjnash, I'm not entirely sure how ownership works in relationship to the GC. I've read the section about GC in the docs but this is literally the first time I've ever directly done anything directly with a garbage collector. Is there any reason why we need to have a token for ownership directly stored with the this buffer type since other types don't require something similar stored with each instance?

**Comment by Tokazama** [comment]
@oscardssmith, I think i figured out how to get GC working for this if you want to give it a try.

**Comment by Tokazama** [comment]
Do note that until "codegen.cpp" has the appropriate LLVM code injected it's going to be pretty slow.

File: base/exports.jl [hunk 1]
```diff
@@ -453,6 +453,7 @@ export
     sum!,
     sum,
     to_indices,
+    unsafe_freeze!,
```
Comment (vchuravy) [inline base/exports.jl:456]: Erroneous export?

Comment (Tokazama) [reply base/exports.jl]: Left over from when `ImmutableBuffer` was part of the PR.

**Comment by oscardssmith** [comment]
I'm going to throw triage on this to see what people think about it.

**Comment by Tokazama** [comment]
Sounds good. I think I got most stuff working this time. I figured that implementing a shared pointer was too much to do in a addition to this so this is mainly following what Array does for now.

I haven't been able to put any serious effort into seeing if we can use the same C type (`jl_buffer_t`) for an additional Julia type that permits dynamic sizing. Looking around at some other PRs (such as https://github.com/JuliaLang/julia/pull/43384), it might be possible to relocate the bits that communicate with the GC outside of the C type.

File: base/summarysize.jl [hunk 2]
```diff
@@ -148,6 +153,27 @@ function (ss::SummarySize)(obj::Array)
     return size
 end
 
+# FIXME Buffer GC
```
Comment (mkitti) [inline base/summarysize.jl:156]: Is this still pending?

Comment (Tokazama) [reply base/summarysize.jl]: I forgot to go back and finish this once I figured out how to get malloced stuff working, but the results are still only as accurate as my assumptions about how jl_buffer_t is working behind the scenes.

**Comment by Tokazama** [comment]
> This is also really making me think of [`java.nio.Buffer`](https://download.java.net/java/early_access/panama/docs/api/java.base/java/nio/Buffer.html) and the more recent [`jdk.incubator.foreign.MemorySegment`](https://download.java.net/java/early_access/loom/docs/api/jdk.incubator.foreign/jdk/incubator/foreign/MemorySegment.html). I wonder if there are any lessons we could learn from that approach.
> 
> A concept that we may want to investigate is that of an [`Arena`](https://download.java.net/java/early_access/jdk20/docs/api/java.base/java/lang/foreign/Arena.html) that controls the lifecycle of these memory segments. Java faces similar issues of garbage collection as Julia does.

One of the reasons I moved away from using shared pointers/references for boxed types here is because we need to have someone come in and standardize the API for memory management in Julia, including GC and pointer/reference types. In an ideal world I could do this all in Julia with something as simple as

```julia
struct Buffer{T}
    length::Int
    data::ReferenceCountedPointer{T}
end
```

The way `Buffer` is implemented in the current PR punts the issue to whoever standardizes the backend down the road.

**Comment by Tokazama** [comment]
@mkitti did you have any suggestions on how the concepts from MemorySegment could be applied here, or is that something that will take more work with the garbage collector first?

**Comment by gbaraldi** [comment]
One thing I was thinking about with this PR is that we might want to make it have a ccall barrier, and move most of the logic either to an LLVM intrinsic or julia itself. Doing that would allow for further constant propagation and might make it easier for LLVM to do things like turning a heap allocation into a stack one.

**Comment by Tokazama** [comment]
> One thing I was thinking about with this PR is that we might want to make it have a ccall barrier, and move most of the logic either to an LLVM intrinsic or julia itself. Doing that would allow for further constant propagation and might make it easier for LLVM to do things like turning a heap allocation into a stack one.

I agree that it would be good to move more of the code to Julia so that optimizations in our compiler implementation also help us in the future.

I believe I'd need to change how everything is allocated to actually get things on stack for this. I figured we'd want to take a more serious attempt at stack allocation when we have the `ImmutableBuffer` type since it fits within the current paradigm of how we treat stack allocated data, but it was pointed out that added some unnecessary complexity to the initial PR so I took it out.

**Comment by Tokazama** [comment]
I've update the OP and made a pretty good amount of progress moving things from C/C++ to Julia.

The main issue I'm running into now is optimizations around pointers. This is pretty well illustrated by comparing the indexing that been optimized by "codegen.cpp" versus just pointer manipulation in native Julia.

```julia
julia> b1 = Buffer{Union{Int, Float64}}(undef, 3);

julia> b2 = Buffer{Int}(undef, 3);

julia> b3 = Buffer{Dict{Int,Int}}(undef, 3);

julia> b3[2] = Dict{Int,Int}();

julia> @btime getindex($b1, 2)
  3.848 ns (0 allocations: 0 bytes)
0.0

julia> @btime Base.get_buffer_value($b1, 2, true, true)
  1.055 μs (6 allocations: 112 bytes)
0.0

julia> @btime getindex($b2, 2)
  4.167 ns (0 allocations: 0 bytes)
4797779568

julia> @btime Base.get_buffer_value($b2, 2, true, true)
  4.460 ns (0 allocations: 0 bytes)
4797779568

julia> @btime getindex($b3, 2)
  5.054 ns (0 allocations: 0 bytes)
Dict{Int64, Int64}()

julia> @btime Base.get_buffer_value($b3, 2, true, true)
  9.512 ns (1 allocation: 80 bytes)
Dict{Int64, Int64}()
```

1. Bits: access to bits types isn't too difficult to optimize. I could probably work on ensuring this
2. Boxed: We have to generate two extra pointers here which may be interfering with inlining.
3. Bits union: this one is really terrible and this may also be suffering from multiple pointers being generated, along with loading extra bits along with looking up the union variant type.

Eventually we'll need a way of also interacting with `jl_gc_wb` when setting new boxed values, but it would probably good to work on avoiding allocations in the getter method first to ensure this overall strategy can work.

**Comment by Tokazama** [comment]
> One thing I was thinking about with this PR is that we might want to make it have a ccall barrier, and move most of the logic either to an LLVM intrinsic or julia itself. Doing that would allow for further constant propagation and might make it easier for LLVM to do things like turning a heap allocation into a stack one.

@gbaraldi, I think you're more familiar with the internal mechanics behind allocation and GC here. Do you have any suggestions concerning what entry points into the GC I should be using to track newly assigned pointers through a `ccall` on the Julia side and facilitating stack allocations?

**Comment by Tokazama** [comment]
Despite some rough edges, I think this is ready for review. I'm still getting some odd errors when resizing `DynamicBuffer`, but besides that all else seems to work fine.

**Comment by mkitti** [comment]
Regarding garbage collection, is anyone here talking to the folks (e.g. @NHDaly et al) working on memory management toolkit.

https://github.com/mmtk/mmtk-julia or https://github.com/mmtk/julia

**Comment by JeffBezanson** [comment]
This is a large and important enough change that I think we need a detailed design document with everything thought through, and with wide buy-in before a lot of code is written. Every detail here can have wide-reaching implications. Some tricky points are how to accommodate non-moving resizing (which becomes very important for large data), and the existing `Array` feature of wrapping foreign pointers. Do we really want both growable and non-growable buffers? We need everybody to be comfortable with the trade-offs chosen here, and compare some alternate designs. @vtjnash, @oscardssmith and I started working on a proposal, but this will take some time.

**Comment by NHDaly** [comment]
Thanks for the ping, @mkitti. CC: @diogo and @kpamnany

**Comment by jpsamaroo** [comment]
:+1: to growable and non-growable buffers (especially in the context of wrapping foreign pointers, where the underlying memory may not support it)

File: src/cgutils.cpp [hunk 3]
```diff
@@ -442,7 +447,8 @@ static size_t dereferenceable_size(jl_value_t *jt)
 // Return the min required / expected alignment of jltype (on the stack or heap)
 static unsigned julia_alignment(jl_value_t *jt)
 {
-    if (jl_is_array_type(jt)) {
+    // TODO BUffer: is it safe to assume will also always have this alignment?
```
Comment (mkitti) [inline src/cgutils.cpp:450]: Todo

**Review by mkitti — COMMENTED** [review]
I highlighted the remaining fixme and Todo comments in the code

**Comment by Tokazama** [comment]
> This is a large and important enough change that I think we need a detailed design document with everything thought through, and with wide buy-in before a lot of code is written. Every detail here can have wide-reaching implications. Some tricky points are how to accommodate non-moving resizing (which becomes very important for large data), and the existing `Array` feature of wrapping foreign pointers. Do we really want both growable and non-growable buffers? We need everybody to be comfortable with the trade-offs chosen here, and compare some alternate designs. @vtjnash, @oscardssmith and I started working on a proposal, but this will take some time.

Is this proposal in progress located anywhere or is the goal to have something more complete put together before getting too many people involved?

**Comment by JeffBezanson** [comment]
Draft proposal in progress here: https://gist.github.com/JeffBezanson/a25dde3bebb5a734af87bb5ddcf31fb0

**Comment by Tokazama** [comment]
I'm going to close this PR since Jeff's proposal already has some clear advantages over this approach. I'd be happy to help further work in this area if it would be helpful going forward. On the other hand, I understand too much feedback at this stage may be counterproductive so I'll stand by until further involvement is solicited.

**Pull request closed without merging**